packages/component-base/src/data-provider-controller.js

```diff
@@ -0,0 +1,118 @@
+import { Cache } from './data-provider-controller/cache.js';
+import { getFlatIndexByPath, getFlatIndexInfo } from './data-provider-controller/helpers.js';
+
+export class DataProviderController extends EventTarget {
+  constructor(host, { size, pageSize, isExpanded, dataProvider, dataProviderParams }) {
+    super();
+    this.host = host;
+    this.size = size;
+    this.pageSize = pageSize;
+    this.isExpanded = isExpanded;
+    this.dataProvider = dataProvider;
+    this.dataProviderParams = dataProviderParams;
+    this.rootCache = this.#createRootCache();
+  }
+
+  get effectiveSize() {
+    return this.rootCache.effectiveSize;
+  }
+
+  isLoading() {
```

Comment:
**note:** I decided not to make it a getter because that complicated overriding it in the [Grid connector](https://github.com/vaadin/flow-components/pull/5239/files#diff-255f48ed31d114d33fd62d5f44b260202e622606b1f6ffb8fa39fbbc96306a84R35-R38).